jax/interpreters/pxla.py

```diff
@@ -71,77 +72,26 @@
 Index = Union[int, slice, Tuple[Union[int, slice], ...]]
 
 
-class NoSharding:
-
-  def __eq__(self, other):
-    return isinstance(other, NoSharding)
-
-  def __repr__(self):
-    return "NoSharding()"
-
+# See the C++ code for comments.
+NoSharding = pmap_lib.NoSharding
+Chunked = pmap_lib.Chunked
+Unstacked = pmap_lib.Unstacked
 
 _UNSHARDED_INSTANCE = NoSharding()
 
-# mypy is very unhappy about taggedtuple
-if TYPE_CHECKING:
-  class Unstacked(NamedTuple):
-    size: int
-else:
-  Unstacked = taggedtuple('Unstacked', ('size',))
 
-class Chunked:
-  chunks: Tuple[int, ...]
+def make_chunked(chunks: Union[int, Sequence[int]]):
+  if isinstance(chunks, int):
+    return Chunked([chunks])
+  return Chunked(list(chunks))
 
-  def __init__(self, chunks: Union[int, Tuple[int, ...]]):
-    if not isinstance(chunks, tuple):
-      chunks = (chunks,)
-    object.__setattr__(self, 'chunks', chunks)
+ShardedAxis = pmap_lib.ShardedAxis
+Replicated = pmap_lib.Replicated
 
-  def __setattr__(self, name, value):
-    raise RuntimeError("Chunked is immutable")
+# mypy cannot deal with the C++ types.
+AvalDimSharding = Union[Unstacked, Chunked, NoSharding]  # type: ignore
+MeshDimAssignment = Union[ShardedAxis, Replicated]  # type: ignore
 
-  def __delattr__(self, name):
-    raise RuntimeError("Chunked is immutable")
-
-  def __hash__(self):
-    return hash(self.chunks)
-
-  def __eq__(self, other):
-    return type(other) is Chunked and self.chunks == other.chunks
-
-  def __repr__(self):
-    return f'Chunked({self.chunks})'
-
-"""
-Represents all the ways we can shard a dimension.
-- `None` means no sharding;
-- `Chunked` means that the dimension is split into the specified number of chunks,
-  but the split dimension itself is preserved inside the map;
-- `Unstacked` means that the dimension is split into chunks of size 1, and doesn't
-  appear inside the map.
-"""
```

Comment:
Lost comment, which is still accurate

Reply:
The comments has been moved to the C++ (it's said on line 75 with
# See the C++ code for comments.
)

See https://github.com/tensorflow/tensorflow/blob/a1f732f7b04a152fc1b2d493cab59a686a0bdb7a/tensorflow/compiler/xla/python/pmap_lib.h#L54

Given the Python only reference the C++ code, and to prevent having to duplicate documentation that increases the chances that one of the 2 versions of the same doc get stale, I think it's a better idea to keep only once place for the comments. I think that, given it's defined in C++, it makes sense to have the doc there.

WDYT?

Reply:
Sounds good to me.